### PR TITLE
Two-level namespace

### LINKED ISSUE #32: Support two-level namespace
After reviewing #31, it seems a piece of missing functionality in fishhook is support for mach-o two-level namespaces. When rebinding a symbol, the logic should check if a specific library has been recorded for that symbol, and if the symbol is bound to a library, then never rebind symbols that are bound to other libraries.

### PR DESCRIPTION
**Pull request by kastiglione**
This change adds support for rebinding that leverages mach-o's two-level namespace.

The tl;dr of two-level namespaces is that symbols imported from other libraries will reference which library they came from. When rebinding a symbol by name alone, it's possible to unintentionally rebind symbols that happen to have the same name, which can cause unexpected results. Using the two-level namespace prevents the mysterious bugs and crashes that can happen when rebinding by just symbol name alone.

In addition to correctness, this change opens a new performance optimization, see 7cc05152e959e16f81ab61fba3b33c872fc687b8.

This change is a breaking change: callers of `rebind_symbols` now need to fill in a `dylib` field on the `rebinding` struct. This field can be `NULL` to opt-out of two-level namespaces, and enable the previous global search.

Resolves #32

### CODE REVIEW

**Comment by kastiglione** [comment]
I've cleaned this up, hopefully this can get a reviewer or two. @grp do you have any thoughts on this, either in general or on the code?

**Comment by grp** [comment]
I don't know much about the implementation of two-level namespacing. :( It looked cool reading through it, though.

**Comment by kastiglione** [comment]
I'll review this in a ~week, hopefully that will give me somewhat fresh eyes to catch bugs, issues, readability problems, etc. I'd like to write a few more tests too ([though I probably won't check the tests in](https://github.com/facebook/fishhook/pull/33#issuecomment-246157967)).
